crates/noirc_evaluator/src/errors.rs

```diff
@@ -8,95 +8,105 @@
-                | InternalError::MissingArg { location, .. }
-                | InternalError::NotAConstant { location, .. }
-                | InternalError::UndeclaredAcirVar { location }
-                | InternalError::UnExpected { location, .. },
+                InternalError::DegreeNotReduced { call_stack }
+                | InternalError::EmptyArray { call_stack }
+                | InternalError::General { call_stack, .. }
+                | InternalError::MissingArg { call_stack, .. }
+                | InternalError::NotAConstant { call_stack, .. }
+                | InternalError::UndeclaredAcirVar { call_stack }
+                | InternalError::UnExpected { call_stack, .. },
             )
-            | RuntimeError::FailedConstraint { location, .. }
-            | RuntimeError::IndexOutOfBounds { location, .. }
-            | RuntimeError::InvalidRangeConstraint { location, .. }
-            | RuntimeError::TypeConversion { location, .. }
-            | RuntimeError::UnInitialized { location, .. }
-            | RuntimeError::UnknownLoopBound { location }
-            | RuntimeError::AssertConstantFailed { location }
-            | RuntimeError::UnsupportedIntegerSize { location, .. } => *location,
+            | RuntimeError::FailedConstraint { call_stack, .. }
+            | RuntimeError::IndexOutOfBounds { call_stack, .. }
+            | RuntimeError::InvalidRangeConstraint { call_stack, .. }
+            | RuntimeError::TypeConversion { call_stack, .. }
+            | RuntimeError::UnInitialized { call_stack, .. }
+            | RuntimeError::UnknownLoopBound { call_stack }
+            | RuntimeError::AssertConstantFailed { call_stack }
+            | RuntimeError::UnsupportedIntegerSize { call_stack, .. } => call_stack,
         }
     }
 }
 
 impl From<RuntimeError> for FileDiagnostic {
-    fn from(error: RuntimeError) -> Self {
-        let file_id = error.location().map(|loc| loc.file).unwrap();
-        FileDiagnostic { file_id, diagnostic: error.into() }
+    fn from(error: RuntimeError) -> FileDiagnostic {
+        let call_stack = vecmap(error.call_stack(), |location| *location);
+        let diagnostic = error.into_diagnostic();
+        let file_id = call_stack.last().map(|location| location.file).unwrap_or_default();
+
+        diagnostic.in_file(file_id).with_call_stack(call_stack)
     }
 }
 
-impl From<RuntimeError> for Diagnostic {
-    fn from(error: RuntimeError) -> Diagnostic {
-        match error {
-            RuntimeError::InternalError(_) => Diagnostic::simple_error(
-                "Internal Consistency Evaluators Errors: \n 
-                This is likely a bug. Consider Opening an issue at https://github.com/noir-lang/noir/issues".to_owned(),
-                "".to_string(),
-                noirc_errors::Span::new(0..0)
-            ),
+impl RuntimeError {
+    fn into_diagnostic(self) -> Diagnostic {
+        match self {
+            RuntimeError::InternalError(_) => {
```

Comment:
Not related: We probably want to get rid of this in another PR and just panic whenever we have an internal Error

Reply:
Hmm I rescind, as I see that it gives us better UX versus the panic trace from Rust